src/Altinn.App.Core/Features/Validation/ValidationService.cs

```diff
@@ -0,0 +1,156 @@
+        var application = await _appMetadata.GetApplicationMetadata();
+        var dataTypesForTask = application.DataTypes.Where(dt => dt.TaskId == taskId).ToList();
+        var dataElementsToValidate = instance.Data.Where(de => dataTypesForTask.Exists(dt => dt.Id == de.DataType)).ToArray();
+        var dataIssuesTask = Task.WhenAll(dataElementsToValidate.Select(dataElement=>ValidateDataElement(instance, dataElement, dataTypesForTask.First(dt=>dt.Id == dataElement.DataType) )));
+
+        return (await Task.WhenAll(taskIssuesTask, dataIssuesTask)).SelectMany(x=>x.SelectMany(y=>y)).ToList();
+    }
+
+
+    /// <inheritdoc/>
+    public async Task<List<ValidationIssue>> ValidateDataElement(Instance instance, DataElement dataElement, DataType dataType)
+    {
+        ArgumentNullException.ThrowIfNull(instance);
+        ArgumentNullException.ThrowIfNull(dataElement);
+        ArgumentNullException.ThrowIfNull(dataElement.DataType);
+
+        // Get both keyed and non-keyed validators for the data type
+        var validators = _serviceProvider.GetServices<IDataElementValidator>()
+            .Where(v => v.DataType == "*" || v.DataType == dataType.Id)
+            // .Concat(_serviceProvider.GetKeyedServices<IDataElementValidator>(dataElement.DataType))
+            .ToArray();
+
+        var dataElementsIssuesTask = Task.WhenAll(validators.Select(async v =>
+        {
+            try
+            {
+                _logger.LogDebug("Start running validator {validatorName} on {dataType} for data element {dataElementId} in instance {instanceId}", v.GetType().Name, dataElement.DataType, dataElement.Id, instance.Id);
+                var issues = await v.ValidateDataElement(instance, dataElement, dataType);
+                issues.ForEach(i => i.Source = v.ValidationSource); // Ensure that the source is set to the validator source
+                return issues;
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Error while running validator {validatorName} on {dataType} for data element {dataElementId} in instance {instanceId}", v.GetType().Name, dataElement.DataType, dataElement.Id, instance.Id);
+                throw;
+            }
+        }));
+
+        // Run extra validation on form data elements with app logic
+        if(dataType.AppLogic?.ClassRef is not null)
+        {
+            Type modelType = _appModel.GetModelType(dataType.AppLogic.ClassRef);
+
+            Guid instanceGuid = Guid.Parse(instance.Id.Split("/")[1]);
+            string app = instance.AppId.Split("/")[1];
+            int instanceOwnerPartyId = int.Parse(instance.InstanceOwner.PartyId);
+            var data = await _dataClient.GetFormData(instanceGuid, modelType, instance.Org, app, instanceOwnerPartyId, Guid.Parse(dataElement.Id)); // TODO: Add method that accepts instance and dataElement
+            var formDataIssuesDictionary = await ValidateFormData(instance, dataElement, dataType, data, null);
```

Comment:
No need to define the default value
```suggestion
            var formDataIssuesDictionary = await ValidateFormData(instance, dataElement, dataType, data);
```